AppDB/appscale/datastore/zkappscale/tornado_kazoo.py

```diff
@@ -1,6 +1,410 @@
+    ConnectionLoss,
+    OperationTimeoutError,
+    ForceRetryError
+  )
+
+  EXPIRED_EXCEPTIONS = (
+    SessionExpiredError,
+  )
+
+  def __init__(self, max_tries=1, delay=0.1, backoff=2, max_jitter=0.8,
+               max_delay=60, ignore_expire=True, deadline=None,
+               interrupt=None):
+    """ Creates an AsyncKazooRetry for retrying coroutines.
+
+    Args:
+      max_tries: How many times to retry the command. -1 means infinite tries.
+      delay: Initial delay between retry attempts.
+      backoff: Backoff multiplier between retry attempts. Defaults to 2 for
+        exponential backoff.
+      max_jitter: Additional max jitter period to wait between retry attempts
+        to avoid slamming the server.
+      max_delay: Maximum delay in seconds, regardless of other backoff
+        settings. Defaults to one minute.
+      ignore_expire: Whether a session expiration should be ignored and treated
+        as a retry-able command.
+      interrupt:
+        Function that will be called with no args that may return
+        True if the retry should be ceased immediately. This will
+        be called no more than every 0.1 seconds during a wait
+        between retries.
+
+    """
+    self.max_tries = max_tries
+    self.delay = delay
+    self.backoff = backoff
+    self.max_jitter = int(max_jitter * 100)
+    self.max_delay = float(max_delay)
+    self._attempts = 0
+    self._cur_delay = delay
+    self.deadline = deadline
+    self._cur_stoptime = None
+    self.retry_exceptions = self.RETRY_EXCEPTIONS
+    self.interrupt = interrupt
+    if ignore_expire:
+      self.retry_exceptions += self.EXPIRED_EXCEPTIONS
+
+  def reset(self):
+    """ Resets the attempt counter. """
+    self._attempts = 0
+    self._cur_delay = self.delay
+    self._cur_stoptime = None
+
+  def copy(self):
+    """ Returns a clone of this retry manager. """
+    obj = AsyncKazooRetry(max_tries=self.max_tries,
+                          delay=self.delay,
+                          backoff=self.backoff,
+                          max_jitter=self.max_jitter / 100.0,
+                          max_delay=self.max_delay,
+                          deadline=self.deadline,
+                          interrupt=self.interrupt)
+    obj.retry_exceptions = self.retry_exceptions
+    return obj
+
+  @gen.coroutine
+  def __call__(self, func, *args, **kwargs):
+    """ Calls a coroutine with arguments until it completes without
+    throwing a Kazoo exception.
+
+    Args:
+      func: Coroutine to yield
+      args: Positional arguments to call the function with
+      kwargs: Keyword arguments to call the function with
+
+    The coroutine will be called until it doesn't throw one of the
+    retryable exceptions (ConnectionLoss, OperationTimeout, or
+    ForceRetryError), and optionally retrying on session
+    expiration.
+    """
+    self.reset()
+
+    while True:
+      try:
+        if self.deadline is not None and self._cur_stoptime is None:
```

Comment:
Yes, I think you're right. However, I'd like to keep the diffs against https://github.com/python-zk/kazoo/blob/c48f2733f2a6b2c2941738e4208e8cfede676730/kazoo/retry.py#L32 as small as possible since comparing the two is the easiest way to see what changes were made to make it async.